### ISSUE TITLE
[Proposal] Remove console.error from js class CrudField constructor

### ISSUE THREAD

**Issue by skpr-repo** [Laravel-Backpack/ideas#200]
I often create a single js file to manage both setupCreateOperation and setupUpdateOperation views. The create and update views can have different fields.

Then I use the crud.field('somefield') to check if the field has been declared or not.

```
if(crud.field('somefield').input) {
  // do something
}
```

The problem is when the field does not exist, the constructor prints a console.error line

`console.error(`CrudField error! Could not select INPUT for "${this.name}"`);`

Even if this is not an error and it's not blocking, it's not elegant and there is always a scared customer complaining "_OMG! There is an error in your code!!! Look at that!!!_"

I know that I could create two different js files, but I would end up duplicating the same functions, because the create and update views share many common fields and behaviours.

You could change the console.error with a less scary console.log or console.info or just remove it.

What do you think about it?

Thanks :-)

**Comment by tabacitu** [comment]
Hmm... interesting. 

That error was thrown there on purpose, to help devs know what exactly has happened wrong. I'm a bit skeptical about removing it... because it would also remove that better developer experience... it would make you _clients_ happier, but it might make _developers_ unhappy... soo... 🤷‍♂️

I'm wondering... @pxpm ... isn't there a workaround for this? Kind of an `field_exists()` statement, that @skpr-repo  can use? Not part of the API, but like... a class he can check for, or something?

Cheers!

**Comment by skpr-repo** [comment]
Well, in my opinion, if we cannot instantiate the CrudField class when the field does not exist, we should raise an error, not an error log.

If the log is just a reminder to help debugging the code, it should be an info or warning, not an error, because technically speaking we are allowed to instantiate the class and check for input existance.

Anyway, a method to check for field existance before instantiating the class would be useful.

Something like:

vendor\backpack\crud\src\resources\views\crud\inc\form_fields_script.blade.php

```
window.crud = {
        ...window.crud,

        // Subfields callbacks holder
        subfieldsCallbacks: [],

        // Create a field from a given name
        field: name => new CrudField(name),

        // Create all fields from a given name list
        fields: names => names.map(window.crud.field),

        **// NEW: Check for field existance
        field_exists: name => $(`[bp-field-name="${name}"]`).length > 0**
};
```

This code does not distinguish fields from subfields, it's just a basic solution, but it works.

Another solution would be to check for input existance and raise a log error inside every public method (show, hide, ...) but it would be time consuming for you... and if the input does not exist, an error would be raised anyway.

**Comment by tabacitu** [comment]
Fair enough. What do you think @pxpm ?

**Comment by pxpm** [comment]
Well, there are a few valid points here:
- You can conditionally load the JS with `@if($crud->hasFieldWhere('name', 'my_field_name')`, but I am not very confortable mixing PHP and JS in a JS file ... it seems meh and it would probably require alot of conditionals in complex scenarios.
- Removing the "error" is a no-go for me, but I would not mind in having it as a `warning`, I still don't like this from the "end user perspective", errors/warnings in the Application are usually a bad sign and transmit bad vibes to the users, I agree with @skpr-repo here
- a `fieldExists` JS function seems like a reasonable solution for people that are "using missing fields on purpose". 

So about the implementation of them:
- The first one is done.

- I think the second one is the "KISS" solution, as a `warning` it would still not be the same as no errors, but atleast it would not show as a **red fat error** (usually those break the app, and the errors we show are not breaking the app). 
I am not 100% convinced about this since to check if the field exists dev need to trigger the error/warning and then check `field.input` presence.

- About the third we will probably need to tinker abit about it ... I am sure we don't want to "repeat the selectors" on the function, so probably we need to introduce a new `CrudField` parameter or something alike: `debug`. 
It can be `true/false`, by default `true`. In the `fieldExists` function we would do `CrudField('name', false)`, and when this `debug` is false we don't show the error message, we return false on error, or the CrudField object when succeeded finding the field.

Probably if we introduce the "debug" parameter to disable the error/warning display we wouldn't need the `fieldExists` function at all since developer could do something like:

```javascript
let slugField = CrudField('slug', false):
// false if it does not exist, or the CrudField object
if (slugField) {
    slugField.onChange(...)
}
```
You know which one I am cheering for. For me, the main benefit is not having to check ``field.input ?? false` after creating the CrudField object. 

Let me know what you think @tabacitu

**Comment by skpr-repo** [comment]
Hello guys, I still miss the benefit of raising the error log during class declaration.

From my perspective as a dev: if I use crud.field('someField').show() and the field does not exist, I get an error, the code breaks, I debug it and I find that the field is missing. So no need to have the error log.

Using a "debug=false" param when instantiating the CrudField class sounds like: "I'm aware that I'm doing something wrong, but I want to bypass the error log". Quite strange.

If your goal is to make the missing field error clearer, you should really catch the error inside every function that uses input or wrapper. This way you could even avoid to break the code if a field is missing or misspelled (not necessarily a benefit).

Or you could create a function that checks if the input or wrapper field is undefined and raises a log error, then insert it at the beginning of every method (show, hide, ecc). This way the code breaks but you have an extra log message. 

But yes, it takes some more time (well... half an hour?).

Anyway it's not a real problem, I'm already checking the field existence with a simple dom check. It's more an aesthetic / usability point of view... so I don't want to bother you more than you need :-)

Thanks!

**Comment by pxpm** [comment]
Well maybe @skpr-repo solution goes in-line with what I purposed without adding the parameter. 

We just move the error raising to the methods, so `CrudField('field_that_does_not_exist')` would not raise the error, but `CrudField('field_that_does_not_exist').show()` would raise it. 

One thing that we should avoid is "letting developer use their dom search", as we know for sure we need to take care for subfields and more complex fields. So .. it's bound to get back to us with difficult/time consuming issues to debug. 

If we move the error to the methods like @skpr-repo suggested we should create a `CrudField('field_name').exists()` as an abstraction for checking `if($crudField.input !== 'undefined')` helping the devs to write cleaner code. 

None of the proposed solutions are time consuming, what consumes time here is thinking about the implications/mantainability and choose the best path to solve this issue. 

After thinking again I think this last solution is the best @tabacitu 

Cheers

**Comment by skpr-repo** [comment]
I totally agree. Maybe it would be useful to take into account even custom fields, especially those without common input types. As an example, since the custom_html field does not let you add closures as values (as opposed to the custom_html column), I created a custom field that shows custom html with a closure. In another case I created a custom field with a button that shows a subset of input fields. Or a hyperlink field for a child item in a subfield.

And even your standard custom_html field has the same problem: even if it exists and it can be managed with js, it gives the usual console log error (only for input, but not for wrapper).

So it would be useful to check for the field existence, not necessarily the input.

Thanks

**Comment by skpr-repo** [comment]
Yes I did it this way, thanks. About the custom_html or custom fields without an input (like mine hyperlink to child items in relationship repeaters), it would be nice to manage them with the same js library, even if they are not input fields, but they can be managed with normal js and dom manipulation, so it does not really matter :-)
Thanks for your help!
